oci/layout/oci_delete.go

```diff
@@ -0,0 +1,245 @@
+func (ref ociReference) getBlobsUsedInIndex(destination map[digest.Digest]int, index *imgspecv1.Index, sharedBlobsDir string) error {
+	for _, descriptor := range index.Manifests {
+		destination[descriptor.Digest]++
+		switch descriptor.MediaType {
+		case imgspecv1.MediaTypeImageManifest:
+			manifest, err := ref.getManifest(&descriptor, sharedBlobsDir)
+			if err != nil {
+				return err
+			}
+			for digest, count := range ref.getBlobsUsedInManifest(manifest) {
+				destination[digest] += count
+			}
+		case imgspecv1.MediaTypeImageIndex:
+			blobPath, err := ref.blobPath(descriptor.Digest, sharedBlobsDir)
+			if err != nil {
+				return err
+			}
+			index, err := parseIndex(blobPath)
+			if err != nil {
+				return err
+			}
+			err = ref.getBlobsUsedInIndex(destination, index, sharedBlobsDir)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unsupported mediaType in index: %q", descriptor.MediaType)
+		}
+	}
+
+	return nil
+}
+
+func (ref ociReference) getBlobsUsedInManifest(manifest *imgspecv1.Manifest) map[digest.Digest]int {
+	blobsUsedInManifest := make(map[digest.Digest]int, 0)
+
+	blobsUsedInManifest[manifest.Config.Digest]++
+	for _, layer := range manifest.Layers {
+		blobsUsedInManifest[layer.Digest]++
+	}
+
+	return blobsUsedInManifest
+}
+
+// This takes in a map of the digest and their usage count in the manifest to be deleted
+// It will compare it to the digest usage in the root index, and return a set of the blobs that can be safely deleted
+func (ref ociReference) getBlobsToDelete(blobsUsedByDescriptorToDelete map[digest.Digest]int, sharedBlobsDir string) (*set.Set[digest.Digest], error) {
+	rootIndex, err := ref.getIndex()
+	if err != nil {
+		return nil, err
+	}
+	blobsUsedInRootIndex := make(map[digest.Digest]int)
+	err = ref.getBlobsUsedInIndex(blobsUsedInRootIndex, rootIndex, sharedBlobsDir)
+	if err != nil {
+		return nil, err
+	}
+
+	blobsToDelete := set.New[digest.Digest]()
+
+	for digest, count := range blobsUsedInRootIndex {
+		if count-blobsUsedByDescriptorToDelete[digest] == 0 {
+			blobsToDelete.Add(digest)
+		}
+	}
+
+	return blobsToDelete, nil
+}
+
+func (ref ociReference) deleteBlobs(blobsToDelete *set.Set[digest.Digest]) error {
+	for _, digest := range blobsToDelete.Values() {
+		blobPath, err := ref.blobPath(digest, "") //Only delete in the local directory, not in the shared blobs path
+		if err != nil {
+			return err
+		}
+		err = deleteBlob(blobPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func deleteBlob(blobPath string) error {
+	logrus.Debug(fmt.Sprintf("Deleting blob at %q", blobPath))
+
+	err := os.Remove(blobPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			logrus.Info(fmt.Sprintf("Blob at %q not found; it was either previously deleted or is in the shared blobs directory", blobPath))
+			return nil
+		} else {
+			return err
+		}
+	} else {
+		return nil
+	}
```

Comment:
(Absolutely non-blocking:
```go
if err != nil && os.IsNotExist(err) {
    logrus…
    return nil // or err = nil
}
return err
```

would do the same thing, and maybe be a bit clearer that this only targets that one special case.)